mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -63,23 +64,121 @@ class Completions(
     case (_: Ident) :: (_: SeqLiteral) :: _ => false
     case _ => true
 
+  /**
+   * It shows the conditions of the code in the cursor position, with
+   * respect to assessing
+   *  1. whether it makes semantic sense to complete
+   *  the position with symbols that are `class`, `trait`,
+   *  `abstract`, `object`, or method
+   *  1. whether it would make syntactic sense to add `[]` `()` or `{}`
+   *  suffixes to the symbols suggested for completion.
+   *
+   *  Note: In the comments `@@` represents the cursor position.
+   *
+   * @param isTypePosition if this is the position for defining a type such as
+   *                       `val a: In@@` or `def i: scala.IndexedSe@@`
+   * @param isNewPosition if it is the position for instantiation after
+   *                      the new keword, as in `new Fil@@`
+   * @param isInstantiationOrMethodCallPos if it is the position for method call or
+   *                                       instantiation as in `object A{ MyCaseClas@@ }`
+   * @param noSquareBracketExists if there is already a square bracket in the position as in
+   *                              `val a: IndexedMa@@[]`
+   */
+  private case class CursorPositionCondition(
+      isTypePosition: Boolean,
+      isNewPosition: Boolean,
+      isInstantiationOrMethodCallPos: Boolean,
+      noSquareBracketExists: Boolean,
+  ):
+
+    def isObjectValidForPos = (!isTypePosition && !isNewPosition)
+
+    /**
+     * classes abd traits are type symbols. They are not suitable for
+     *        instantiation or method call positions. Only apply methods or objects
+     *        can be used in such positions.
+     */
+    def isClassOrTraitValidForPos = !isInstantiationOrMethodCallPos
+    def isMethodValidForPos = !isTypePosition
+
+    def isParanethesisValidForPos = !isTypePosition
+    def isSquareBracketsValidForPos = noSquareBracketExists
+      && (isTypePosition || isNewPosition || isInstantiationOrMethodCallPos)
+
+    /**
+     * as in ```new MyTrait@@```
+     */
+    def isCurlyBracesValidForPos = isNewPosition
+  end CursorPositionCondition
+
+  private lazy val cursorPositionCondition =
+    calculateTypeInstanceAndNewPositions(path)
+
+  private def calculateTypeInstanceAndNewPositions(
+      path: List[Tree]
+  ): CursorPositionCondition =
+    path match
+      case (head: (Select | Ident)) :: tail =>
+        // this method of calculating hasNoSquareBracket instead of using the
+        // path for the purpose is due to the issue that trees after typer
+        // lose information in dotty:
+        // https://github.com/lampepfl/dotty/issues/15750
```

Comment:
sorry for the confusion, I think I forgot the punctuation. Does it read better with the comma:

```
 // this method of calculating hasNoSquareBracket, instead of using the
 // path for this purpose, is due to the issue that trees after typer
 // lose information in dotty:
```

Reply:
Yeah, now it's better. Although I'd just leave sth like that
`due to the issue in dotty, because of which trees after typer lose information, we have to calculate hasNoSquareBracket manually
`